Sync up with ES main to fix the passing whitelists of Painless.

### Description

Recent Elasticsearch source changes of [Painless whitelists arrangement](https://github.com/elastic/elasticsearch/pull/106913) broke the plugin, see BK CI: https://buildkite.com/elastic/logstash-filter-elastic-integration-build/builds/138
The error we get:
```
48.3 > Task :compileJava
548.3 /usr/share/plugins/plugin/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java:312: error: cannot find symbol
548.3                 IngestScript.CONTEXT, PainlessPlugin.BASE_WHITELISTS,
548.3                                                     ^
548.3   symbol:   variable BASE_WHITELISTS
548.3   location: class PainlessPlugin
548.3 /usr/share/plugins/plugin/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java:313: error: cannot find symbol
548.3                 IngestConditionalScript.CONTEXT, PainlessPlugin.BASE_WHITELISTS);
548.3                                                                ^
548.3   symbol:   variable BASE_WHITELISTS
548.3   location: class PainlessPlugin
548.5 2 errors
548.6
548.6 > Task :compileJava FAILED
548.6
548.6 FAILURE: Build failed with an exception.
548.6
```

Now, Elasticsearch eagerly loads the base whitelists (`PainlessPlugin.baseWhiteList()`) and I have just tried with [draft PR](https://github.com/elastic/logstash-filter-elastic_integration/pull/134) if we have an access for the `baseWhiteList()` but it seems we do not have. So it seems this requires again [`LogstashInternalBridge.java`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/ingest/LogstashInternalBridge.java) to access.

```
646.1 > Task :compileJava
646.1 /usr/share/plugins/plugin/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java:312: error: cannot find symbol
646.1                 IngestScript.CONTEXT, PainlessPlugin.baseWhiteList(),
646.1                                                     ^
646.1   symbol:   method baseWhiteList()
646.1   location: class PainlessPlugin
646.1 /usr/share/plugins/plugin/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java:313: error: cannot find symbol
646.1                 IngestConditionalScript.CONTEXT, PainlessPlugin.baseWhiteList());
646.1                                                                ^
646.1   symbol:   method baseWhiteList()
646.1   location: class PainlessPlugin
646.4 2 errors
646.5
646.5 > Task :compileJava FAILED
```

--- a/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java
+++ b/src/main/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilder.java
@@ -308,13 +308,36 @@ public class EventProcessorBuilder {
     }
 
     private static ScriptService initScriptService(final Settings settings, final ThreadPool threadPool) {
+        final List<Whitelist> painlessBaseWhitelist = getPainlessBaseWhiteList();
         final Map<ScriptContext<?>, List<Whitelist>> scriptContexts = Map.of(
-                IngestScript.CONTEXT, PainlessPlugin.BASE_WHITELISTS,
-                IngestConditionalScript.CONTEXT, PainlessPlugin.BASE_WHITELISTS);
+                IngestScript.CONTEXT, painlessBaseWhitelist,
+                IngestConditionalScript.CONTEXT, painlessBaseWhitelist);
 
         Map<String, ScriptEngine> engines = new HashMap<>();
         engines.put(PainlessScriptEngine.NAME, new PainlessScriptEngine(settings, scriptContexts));
         engines.put(MustacheScriptEngine.NAME, new MustacheScriptEngine());
         return new ScriptService(settings, engines, ScriptModule.CORE_CONTEXTS, threadPool::absoluteTimeInMillis);
+    }
+
+    /**
+     * @implNote handles breaking changes introduced in Elasticsearch 8.14 series; once 8.14 is
+     *           released and all builds of this plugin depend on Elasticsearch 8.14+, this can
+     *           be simplified to call {@code PainlessPlugin.baseWhiteList()} directly.
+     * @return the PainlessPlugin's default base whitelists
+     */
+    @SuppressWarnings({"JavaReflectionMemberAccess", "unchecked"})
+    static List<Whitelist> getPainlessBaseWhiteList() {
+        Class<PainlessPlugin> cls = PainlessPlugin.class;
+        try {
+            try {
+                // In 8.14+: PainlessPlugin.baseWhiteList()
+                return (List<Whitelist>) cls.getMethod("baseWhiteList").invoke(null);
+            } catch (NoSuchMethodException e) {
+                // in 8.x->8.13.x: PainlessPlugin.BASE_WHITELISTS
+                return (List<Whitelist>) cls.getField("BASE_WHITELISTS").get(null);
+            }
+        } catch (java.lang.reflect.InvocationTargetException | IllegalAccessException | NoSuchFieldException e) {
+            throw new RuntimeException("Unsupported PainlessPlugin does not provide access to its base whitelists", e);
+        }
     }
 }

--- a/src/test/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilderTest.java
+++ b/src/test/java/co/elastic/logstash/filters/elasticintegration/EventProcessorBuilderTest.java
@@ -1,0 +1,19 @@
+package co.elastic.logstash.filters.elasticintegration;
+
+import org.elasticsearch.painless.spi.Whitelist;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+
+class EventProcessorBuilderTest {
+
+    @Test
+    void getPainlessBaseWhiteList() {
+        final List<Whitelist> painlessBaseWhiteList = EventProcessorBuilder.getPainlessBaseWhiteList();
+        assertThat(painlessBaseWhiteList, is(not(empty())));
+    }
+}

